### DISCUSSION TITLE
enh: Better CJK language support when downloading pdf

### DISCUSSION THREAD

**Discussion by y0umu** [open-webui/open-webui#4738] [General]
**Is your feature request related to a problem? Please describe.**
When I try to download the conversation as pdf, Chinese (traditional or simplified) charactors are missing glyphs. Examples below

![traditional_chinese](https://github.com/user-attachments/assets/b83b4163-4901-4f7e-94b4-0721bb86581b)
Got `WARNI [fpdf.output] Font MPDFAA+NotoSans is missing the following glyphs: 啟, 鸠` when downloading the pdf.

![simplified_chinese](https://github.com/user-attachments/assets/e2e9f2c8-1e81-4a9f-bce5-b68db207916e)

Got `WARNI [fpdf.output] Font MPDFAA+NotoSans is missing the following glyphs: 谁, 这, 话, 时, 议, 颁, 夺, 财, 产, 为, 发, 许, 宪, 现` when downloading the pdf.

**Describe the solution you'd like**
I workarounded by hacking codes near https://github.com/open-webui/open-webui/blob/99d10d1189452ad49fcace219e9c90ae65906cd1/backend/apps/webui/routers/utils.py#L83 to something like:

```python
    pdf.add_font("NotoSans", "", f"{FONTS_DIR}/NotoSans-Regular.ttf")
    pdf.add_font("NotoSans", "b", f"{FONTS_DIR}/NotoSans-Bold.ttf")
    pdf.add_font("NotoSans", "i", f"{FONTS_DIR}/NotoSans-Italic.ttf")
    pdf.add_font("NotoSansSC", "", f"{FONTS_DIR}/NotoSansSC-Regular.ttf")
    pdf.add_font("NotoSansKR", "", f"{FONTS_DIR}/NotoSansKR-Regular.ttf")
    pdf.add_font("NotoSansJP", "", f"{FONTS_DIR}/NotoSansJP-Regular.ttf")

    pdf.set_font("NotoSans", size=12)
    # TODO select the best fallback sequence by determining the user's locale
    pdf.set_fallback_fonts(["NotoSansSC", "NotoSansKR", "NotoSansJP"])
```
And I have put NotoSansSC-Regular.ttf from https://fonts.google.com/noto into `backend/static/fonts`.

This might not be accepted as the general solution however, as it probably breaks displays of  some Japanese chars (门、入、内 for example)

**Describe alternatives you've considered**
Determine the user's locale before executing `pdf.set_fallback_fonts(["NotoSansSC", "NotoSansKR", "NotoSansJP"])` to make sure the fallback sequence is correct.

**Additional context**
No additionall context

**Comment by tjbck** [comment]
PR welcome!

**Comment by y0umu** [comment]
I did't propose a pull request in the first place because I cannot come to a conclusion on what is the most suitable approach to fetch the locale.

- The locale of backend is the easy to get, but it does not make sense if it is a public server with plain `en_US` or `C` locale .
- Inferring locale from the text fed to the FPDF module might work, but inferring itself sounds tedious.
- Inferring from the WebUI language setting is a good idea, but I haven't found related code at this point.
- Inferring from the request header is also a good idea, but `download_chat_as_pdf` does not seem to have access to request headers at all.

**Comment by tjbck** [comment]
what would be the issue of setting it to `["NotoSansKR", "NotoSansJP", "NotoSansSC"]` in this order?

**Reply by y0umu** [reply]
TL;DR It will break the pdf rendering if something exists in either `["NotoSansKR", "NotoSansJP", "NotoSansSC"]` sharing the same Unicode code point.

Providing a vivid example is quite difficult, as I had to find something both Chinese and Japanese people know. I tested with some ancient Chinese poem (杜牧 《山行》, [Simplified Chinese version](https://www.gushiwen.cn/shiwenv_3c36881bd247.aspx), [Japanese version](https://ja.wikibooks.org/wiki/%E9%AB%98%E7%AD%89%E5%AD%A6%E6%A0%A1%E5%9B%BD%E8%AA%9E%E7%B7%8F%E5%90%88/%E6%BC%A2%E6%96%87/%E5%B1%B1%E8%A1%8C)) and will provide a working example in the next post.

**Comment by y0umu** [comment]
I rewrite `download_chat_as_pdf` as a test function. Code snippet below:
```python
from fpdf import FPDF

def pdf_test(text: str, fallback_seq: list, outname: str):
    pdf = FPDF()
    pdf.add_page()

    STATIC_DIR = "./static"
    FONTS_DIR = f"{STATIC_DIR}/fonts"

    pdf.add_font("NotoSans", "", f"{FONTS_DIR}/NotoSans-Regular.ttf")
    pdf.add_font("NotoSans", "b", f"{FONTS_DIR}/NotoSans-Bold.ttf")
    pdf.add_font("NotoSans", "i", f"{FONTS_DIR}/NotoSans-Italic.ttf")
    pdf.add_font("NotoSansSC", "", f"{FONTS_DIR}/NotoSansSC-Regular.ttf")
    pdf.add_font("NotoSansKR", "", f"{FONTS_DIR}/NotoSansKR-Regular.ttf")
    pdf.add_font("NotoSansJP", "", f"{FONTS_DIR}/NotoSansJP-Regular.ttf")

    pdf.set_font("NotoSans", size=12)
    # TODO select the best fallback sequence by determining the user's locale
    pdf.set_fallback_fonts(fallback_seq)

    pdf.set_auto_page_break(auto=True, margin=15)

    # Adjust the effective page width for multi_cell
    effective_page_width = (
        pdf.w - 2 * pdf.l_margin - 10
    )  # Subtracted an additional 10 for extra padding

    content = text
    pdf.set_font("NotoSans", size=16)  # Regular for content
    pdf.multi_cell(effective_page_width, 6, content, 0, "L")
    pdf.ln(1.5)  # Extra space between messages

    # Save the pdf with name .pdf
    pdf_bytes = pdf.output(name=outname)

if __name__ == "__main__":
    test_sc = """远上寒山石径斜
白云生处有人家
停车坐爱枫林晚
霜叶红于二月花 """
    test_jp = """遠上寒山石径斜
白雲生処有人家
停車坐愛楓林晩
霜葉紅於二月花"""
    pdf_test(test_sc, fallback_seq=["NotoSansSC", "NotoSansJP"], outname='sc_sc-jp_ok.pdf')   # correct sequence
    pdf_test(test_jp, fallback_seq=["NotoSansJP", "NotoSansSC"], outname='jp_jp-sc_ok.pdf')   # correct sequence
    pdf_test(test_sc, fallback_seq=["NotoSansJP", "NotoSansSC"], outname='sc_jp-sc_bad.pdf')   # wrong sequence
    pdf_test(test_jp, fallback_seq=["NotoSansSC", "NotoSansJP"], outname='jp_sc-jp_bad.pdf')   # wrong sequence
```

Now let's see what are the rendered pdf, pay special attention to the chars I bound with red boxes.
![sample](https://github.com/user-attachments/assets/b3dec437-850d-48de-ba89-59562cebb77d)

As you can see, glyphs of "径" in Simplified Chinese and Japanese (or Traditional Chinese, basically Kanji borrows more Traditional Chinese) are completely two things, even if they share the same Unicode. Choosing the wrong fallback sequence is the same as choosing the wrong glyph. There are also slight difference between "家" and "花" as well.

**Comment by y0umu** [comment]
The major problem with CJK rendering is there could be different glyphs for the very same Unicode point in different locales. In the previous post that is '径' with Unicode 0x5f84. This char exists in either Simplified Chinese, Traditional Chinese or Japanese.
